node/src/components/consensus/highway_core/state.rs

```diff
@@ -915,43 +924,41 @@ impl<C: Context> State<C> {
 
     /// Returns whether the unit with `hash0` sees the one with `hash1` (i.e. `hash0 ≥ hash1`),
     /// and sees `hash1`'s creator as correct.
-    fn _sees_correct(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
-        hash0 == hash1 || self.unit(hash0).panorama._sees_correct(self, hash1)
+    pub(crate) fn sees_correct(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
+        hash0 == hash1 || self.unit(hash0).panorama.sees_correct(self, hash1)
     }
 
     /// Returns whether the unit with `hash0` sees the one with `hash1` (i.e. `hash0 ≥ hash1`).
-    fn _sees(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
-        hash0 == hash1 || self.unit(hash0).panorama._sees(self, hash1)
+    pub(crate) fn sees(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
+        hash0 == hash1 || self.unit(hash0).panorama.sees(self, hash1)
     }
 
     // Returns whether the units with `hash0` and `hash1` see each other or are equal.
-    fn _is_compatible(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
-        hash0 == hash1
-            || self.unit(hash0).panorama._sees(self, hash1)
-            || self.unit(hash1).panorama._sees(self, hash0)
+    fn is_compatible(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
+        hash0 == hash1 || self.unit(hash0).panorama.sees(self, hash1)
     }
 
     /// Returns the panorama of the confirmation for the leader unit `vhash`.
+    ///
+    /// Returns `None` if we are not supposed to confirm the proposal.
     pub(crate) fn confirmation_panorama(
         &self,
         own_idx: ValidatorIndex,
-        vhash: &C::Hash,
-    ) -> Panorama<C> {
-        let unit = self.unit(vhash);
-        let mut panorama;
-        // TODO(HWY-167): Confirmation panorama.
-        if let Some(prev_hash) = self.panorama().get(own_idx).correct().cloned() {
-            let own_unit = self.unit(&prev_hash);
-            panorama = unit.panorama.merge(self, &own_unit.panorama);
-            panorama[own_idx] = Observation::Correct(prev_hash);
-        } else {
-            panorama = unit.panorama.clone();
+        uhash: &C::Hash,
+    ) -> Option<Panorama<C>> {
+        let proposal = self.unit(uhash);
+        if let Some(prev_hash) = self.citable_panorama().get(own_idx).correct() {
```

Comment:
I'm a bit worried that we're not using `panorama()` here: If that differs from `citable_panorama` in our _own_ vote, we would create an equivocation here! Since that's much worse than an invalid vote, I'd go with `panorama`.

But actually it should never happen, right? They can never differ in `own_idx`? At least I believe we must make sure it can't happen, otherwise we'd equivocate in our next witness vote or proposal?

To do this, do we actually need to check here whether `uhash` is seen by `citable_panorama` (_in addition_ to whether it sees our latest vote)?
Because imagine we added one of Bob's forks to `citable_panorama` already, and now the proposal comes in and naively cites another fork by Bob (which is fine, because it doesn't cite the first one). We could create a valid confirmation of that, but then we'd actually fail to add our own new confirmation vote to `citable_panorama` because it can't be cited _together_ with Bob's first fork.

Reply:
Since this is dangerous, it would actually be good to have another unit test for that.
So:
* Carol creates `c0`.
* Bob equivocates: `b0`, `b0_prime`.
* Verify that only `b0` has been added to `citable_panorama`, but not `b0_prime`.
* Alice creates a proposal `a0` citing only `b0_prime`.
* Verify that `a0` is not in `citable_panorama`.
* Call `confirmation_panorama(CAROL, a0)` and verify that it returns `None`.
* Endorse `a0`.
* Call `confirmation_panorama(CAROL, a0)` and verify that it returns `a0`'s inclusive panorama.

Reply:
> Endorse `a0`.

And actually, that should even work if it was `b0` or `b0_prime` instead of `a0`, in theory.
Depending on our strategy for creating new endorsements, that is or isn't a requirement… I'm so confused now. :dizzy_face:

Reply:
Done in [`56c543a` (#572)](https://github.com/CasperLabs/casper-node/pull/572/commits/56c543ab37bba33495c4871f21e4b731e1eac277)

Reply:
That isn't done yet, is it?